### PR TITLE
Feature/image mesh

### PR DESCRIPTION
**Pull request by projectitis**
This is a draft PR to show my work in progress for image mesh support.
Any feedback and suggestions welcome.
Relates to #1218 

It is currently implemented as a feature of the Picture object:
```
auto picture = Picture::gen();
picture->load("texture.png");
Polygon triangles[4] = { /* vertex data here (points + UVs) */ }
picture->mesh(&triangles, 4); // Define a mesh made of 4 triangles
canvas->push(move(picture));
```

Note: This branch also contains various other fixes, like the PNG channel-order bug, and better windows support.

### CODE REVIEW

**Comment by projectitis** [comment]
Hi @hermet @JSUYA - 

I have this working now, but there are currently two issues.
1) There is a seam between triangles
2) The clipping is not happening correctly

If you have any idea about how to fix these issues, or where they might be occurring, please let me know where to look!

Here is a video showing the issue:
https://user-images.githubusercontent.com/32183808/167791956-cc03826b-e9b2-4a2e-b46f-31b95fdb46c5.mp4

The code to create this is:
```
auto picture = Picture::gen();
picture->load("texture.png"); // 200 x 200 image (if different, change the UVs in the triangle data)
Polygon triangles[2];
triangles[0] = {{
	{
		{0,0}, // pt
		{0,0} // uv
	},
	{
		{200,-50},
		{200,0}
	},
	{
		{0,200},
		{0,200}
	}
}};
triangles[1] = { {
	{
		{200,-50},
		{200,0}
	},
	{
		{0,200},
		{0,200}
	},
	{
		{200,250},
		{200,200}
	}
} };
picture->mesh(&triangles, 2);
canvas->push(move(picture));
```

**Comment by projectitis** [comment]
Hi @hermet @JSUYA - I have fixed (removed) the seams between the triangles by combining all triangles into one AA span area instead of treating each triangle as it's own one. I suspect that this will cause slower rendering for some shaped meshes because the area might contain many transparent/undrawn pixels. It's better to have no seams, though.

However, still the following issues.
1) The clipping area is still wrong
2) While rotating the mesh, at a certain point an exception is thrown trying to free the AA spans:
```
    free(aaSpans->lines);
    free(aaSpans); // exception thrown on this line
```
I believe that they are both related to each other because it always happens when the clipping area shrinks to it's smallest size.

Demo:
https://user-images.githubusercontent.com/32183808/167963654-0551f13e-70b2-401a-8614-d9f482dc06c0.mp4

**Comment by projectitis** [comment]
`imagePrepare ---> _genOutline` is the problem with the clipping area.
But it does not fix the exception that occurs.

**Comment by projectitis** [comment]
`imagePrepare ---> _genOutline` is the problem with the clipping area.
It was still assuming a rectangular clipping area of a standard image.
I have put in a quick fix (that probably can be improved a lot).
But fixing that does not fix the exception.

**Comment by hermet** [comment]
> Hi @hermet @JSUYA -
> 
> I have this working now, but there are currently two issues.
> 
> 1. There is a seam between triangles
> 2. The clipping is not happening correctly
> 
> If you have any idea about how to fix these issues, or where they might be occurring, please let me know where to look!
> 
> Here is a video showing the issue: https://user-images.githubusercontent.com/32183808/167791956-cc03826b-e9b2-4a2e-b46f-31b95fdb46c5.mp4
> 
> The code to create this is:
> 
> ```
> auto picture = Picture::gen();
> picture->load("texture.png"); // 200 x 200 image (if different, change the UVs in the triangle data)
> Polygon triangles[2];
> triangles[0] = {{
> 	{
> 		{0,0}, // pt
> 		{0,0} // uv
> 	},
> 	{
> 		{200,-50},
> 		{200,0}
> 	},
> 	{
> 		{0,200},
> 		{0,200}
> 	}
> }};
> triangles[1] = { {
> 	{
> 		{200,-50},
> 		{200,0}
> 	},
> 	{
> 		{0,200},
> 		{0,200}
> 	},
> 	{
> 		{200,250},
> 		{200,200}
> 	}
> } };
> picture->mesh(&triangles, 2);
> canvas->push(move(picture));
> ```


Current AA logic is only adaptable to one triangle, it never consider the mesh types. However, you gonna make it, very impressive. :)

Clipping wise, I believe you can figure out and make it work. I believe it won't be difficult because its basic features are already working there. Otherwise you can ignore the clipping feature . We can revisit and enhance it step by step later. Please try to do best, I will revisit you code soon and point out the issue you need helps.  👍

**Comment by projectitis** [comment]
@hermet I am back on this. Made some progress today. This is an image mesh made in the latest version of the Rive editor and rendered using the Rive-Tizen viewer with image mesh implemented!

p.s. That is my new puppy, Ginny.

image

image

**Comment by hermet** [comment]
Good job, looks impressive. Please tell when the patch is ready. I will help for this. Once this patch is applied, I believe you can drive the feature development with dev-access. Also I wanted to talk about Derive SDK you're probably working for. It would be helpful for both of us If you describe your project mission or vision. :-)

**Comment by projectitis** [comment]
@hermet thank you. It is actually the same code as now (no changes).  I am trying to fix the failing test. Then I will re-open the PR.

There is a bit of detail in the Derive SDK description - what information specifically should I add to that?

**Comment by projectitis** [comment]
It would be great if somebody could help with a failing test :) 
* The tests on `master` all pass
* The picture tests on this branch have some failing tests (trying to access a NULL matrix)

But I cannot see the difference between the two branches that causes this 🤔 

This is the test that fails:
```
// testPicture.cppm line 325
TEST_CASE("Load JPG file and render", "[tvgPicture]")
{
    REQUIRE(Initializer::init(CanvasEngine::Sw, 0) == Result::Success);

    auto canvas = SwCanvas::gen();
    REQUIRE(canvas);

    uint32_t buffer[100*100];
    REQUIRE(canvas->target(buffer, 100, 100, 100, SwCanvas::Colorspace::ABGR8888) == Result::Success);

    auto picture = Picture::gen();
    REQUIRE(picture);

    REQUIRE(picture->load(TEST_DIR"/test.jpg") == Result::Success);

    REQUIRE(canvas->push(move(picture)) == Result::Success); // XXXXXXXXXXXXXXXXX Exception here

    REQUIRE(Initializer::term(CanvasEngine::Sw) == Result::Success);
}
```

This is the test result:
```
../test/testPicture.cpp(338): FAILED:
  {Unknown expression after the reported line}
due to a fatal error condition:
  SIGSEGV - Segmentation violation signal
```

This is the call stack when the exception occurs:
```
tvgUnitTests.exe!_onlyShifted(const tvg::Matrix * m) Line 31
tvgUnitTests.exe!imagePrepare(SwImage * image, tvg::Polygon * triangles, unsigned int count, const tvg::Matrix * transform, const SwBBox & clipRegion, SwBBox & renderRegion, SwMpool * mpool, unsigned int tid) Line 123
tvgUnitTests.exe!SwImageTask::run(unsigned int tid) Line 196
tvgUnitTests.exe!tvg::TaskSchedulerImpl::request(tvg::Task * task) Line 156
tvgUnitTests.exe!tvg::TaskScheduler::request(tvg::Task * task) Line 185
tvgUnitTests.exe!tvg::SwRenderer::prepareCommon(SwTask * task, const tvg::RenderTransform * transform, unsigned int opacity, const tvg::Array<void *> & clips, tvg::RenderUpdateFlag flags) Line 623
tvgUnitTests.exe!tvg::SwRenderer::prepare(tvg::Surface * image, tvg::Polygon * triangles, unsigned int triangleCnt, void * data, const tvg::RenderTransform * transform, unsigned int opacity, tvg::Array<void *> & clips, tvg::RenderUpdateFlag flags) Line 640
tvgUnitTests.exe!tvg::Picture::Impl::update(tvg::RenderMethod & renderer, const tvg::RenderTransform * pTransform, unsigned int opacity, tvg::Array<void *> & clips, tvg::RenderUpdateFlag pFlag) Line 134
tvgUnitTests.exe!tvg::PaintMethod<tvg::Picture::Impl>::update(tvg::RenderMethod & renderer, const tvg::RenderTransform * transform, unsigned int opacity, tvg::Array<void *> & clips, tvg::RenderUpdateFlag renderFlag) Line 180
tvgUnitTests.exe!tvg::Paint::Impl::update(tvg::RenderMethod & renderer, const tvg::RenderTransform * pTransform, unsigned int opacity, tvg::Array<void *> & clips, unsigned int pFlag) Line 242
tvgUnitTests.exe!tvg::Canvas::Impl::update(tvg::Paint * paint, bool force) Line 97
tvgUnitTests.exe!tvg::Canvas::Impl::push(std::unique_ptr<tvg::Paint,std::default_delete<tvg::Paint>> paint) Line 57
tvgUnitTests.exe!tvg::Canvas::push(std::unique_ptr<tvg::Paint,std::default_delete<tvg::Paint>> paint) Line 48
```

The exception occurs in `_onlyShifted` because the matrix is NULL.
```
static inline bool _onlyShifted(const Matrix* m)
{
    if (mathEqual(m->e11, 1.0f) && mathEqual(m->e22, 1.0f) && mathZero(m->e12) && mathZero(m->e21)) return true; // <<< m is NULL
    return false;
}
```

Any ideas?

**Comment by projectitis** [comment]
@hermet Ready for review!

File: inc/thorvg.h
Comment (JSUYA) [inline inc/thorvg.h:1218]: How about keeping only appendMesh?
According to the description of appendMesh, there is no need to use this.
@hermet, what do you think?

File: inc/thorvg.h
Comment (JSUYA) [inline inc/thorvg.h:1232]: Among the image types provided by Picture, tvg, svg are not currently supported. Could you please add an explanation for this?

File: inc/thorvg.h
Comment (JSUYA) [inline inc/thorvg.h:1230]: Looking at the code, it seems that `Result::NonSupport` does not returned.
Could you please add an explanation for returning `Result::InsufficientCondition`?

File: src/lib/tvgPictureImpl.h
Comment (JSUYA) [inline src/lib/tvgPictureImpl.h:269]: Could you please add space?
`i<count` -> `i < count`

File: src/loaders/png/tvgLodePng.cpp
Comment (JSUYA) [inline src/loaders/png/tvgLodePng.cpp:2089]: When this change takes effect, https://github.com/Samsung/thorvg/issues/1007#issuecomment-964153604 occurs again. Could you please check?

(https://github.com/Samsung/thorvg/issues/1007#issue-1046668771 Please check the diagram box color on the right of the SVG file attached here)

File: src/lib/tvgPictureImpl.h
Comment (JSUYA) [inline src/lib/tvgPictureImpl.h:269]: Please remove warning message
```
../src/lib/tvgPictureImpl.h: In member function ‘bool tvg::Picture::Impl::mesh(const tvg::Polygon*, uint32_t, bool)’:
../src/lib/tvgPictureImpl.h:269:28: warning: comparison of integer expressions of different signedness: ‘int’ and ‘const uint32_t’ {aka ‘const unsigned int’} [-Wsign-compare]
  269 |             for (int i=0; i<count; i++) this->triangles[this->triangleCnt + i] = triangles[i];
      |                           ~^~~~~~
../src/lib/tvgPictureImpl.h:275:28: warning: comparison of integer expressions of different signedness: ‘int’ and ‘const uint32_t’ {aka ‘const unsigned int’} [-Wsign-compare]
  275 |             for (int i=0; i<count; i++) this->triangles[i] = triangles[i];
```

**Review by JSUYA — COMMENTED** [review]
wow awesome work. Rive example is very impressive. Thank you for your work.
I left a few brief comments. Could you please check?

(I'm so glad we're back in thorvg after a long time )

File: inc/thorvg.h
Comment (projectitis) [reply inc/thorvg.h:1218]: I will change this to a single method that matches the Impl:

`Result mesh(const Polygon* triangles, const uint32_t count, bool append = false) noexcept;`

Is that ok?

File: inc/thorvg.h
Comment (projectitis) [reply inc/thorvg.h:1230]: Thank you for picking this up! I will change this to `Result::InvalidArguments` and add an explanation.

File: src/loaders/png/tvgLodePng.cpp
Comment (projectitis) [reply src/loaders/png/tvgLodePng.cpp:2089]: I'll revert this change in this branch, and we can work on the channel order bug in the other branch.

File: inc/thorvg.h
Comment (projectitis) [reply inc/thorvg.h]: Done

File: inc/thorvg.h
Comment (projectitis) [reply inc/thorvg.h:1226]: Done

File: inc/thorvg.h
Comment (projectitis) [reply inc/thorvg.h]: Done

File: src/lib/tvgPictureImpl.h
Comment (projectitis) [reply src/lib/tvgPictureImpl.h]: Done

File: src/loaders/png/tvgLodePng.cpp
Comment (projectitis) [reply src/loaders/png/tvgLodePng.cpp]: Reverted

File: src/lib/tvgPictureImpl.h
Comment (projectitis) [reply src/lib/tvgPictureImpl.h]: Done

**Comment by projectitis** [comment]
@JSUYA Thank you for the review. I have made the suggested changes. Please do let me know if you have other suggestions.

I did not know if it is your policy for the reviewer to resolve the threads, or for the contributor to do so. I have left them for your to resolve. Please let me know.

File: inc/thorvg.h
Comment (hermet) [inline inc/thorvg.h:191]: Could you please add ?  Once the feature is fully stable we will add the release tag.
@BETA_API

File: inc/thorvg.h
Comment (hermet) [inline inc/thorvg.h:203]: @BETA_API

File: inc/thorvg.h
Comment (hermet) [reply inc/thorvg.h:191]: In my opinion, the uv range would be better to use in the normalized range like (0 ~ 1). 
Otherwise, the user must get the size of the image first. We can skip that usage if it's possible.

File: inc/thorvg.h
Comment (hermet) [inline inc/thorvg.h:1229]: @BETA_API

File: inc/thorvg.h
Comment (hermet) [inline inc/thorvg.h:1240]: @BETA_API

File: inc/thorvg.h
Comment (hermet) [inline inc/thorvg.h:1221]: Do it really need the append function?? I'm worrying if it increases the feature-complexity without any practical benefits.

File: inc/thorvg.h
Comment (hermet) [inline inc/thorvg.h:1224]: I wonder if the user wants to remove the mesh from the Picture instance.  For instance they can remove the existing mesh by setting Polygon = nullptr with count = 0;

File: src/lib/sw_engine/tvgSwImage.cpp
Comment (hermet) [inline src/lib/sw_engine/tvgSwImage.cpp:89]: coding convention
if () {
  ...
} else {
  ...
}

File: src/lib/sw_engine/tvgSwCommon.h
Comment (hermet) [inline src/lib/sw_engine/tvgSwCommon.h:324]: suggest more clear name:
i.e) count -> trianglesCnt
Other parameter names in functions as well.

File: src/lib/sw_engine/tvgSwRasterTexmap.h
Comment (hermet) [inline src/lib/sw_engine/tvgSwRasterTexmap.h:611]: suggestion: Polygon* => auto

File: src/lib/sw_engine/tvgSwRasterTexmap.h
Comment (hermet) [inline src/lib/sw_engine/tvgSwRasterTexmap.h:630]: fix convention
i=0   =>   i = 0
i<count   =>   i < count

File: src/lib/sw_engine/tvgSwRasterTexmap.h
Comment (hermet) [inline src/lib/sw_engine/tvgSwRasterTexmap.h:631]: Probably we can avoid the data copy?
_rasterPolygonImage(surface, image, region, opacity, transformedTris[i], blendMethod, aaSpans);

File: src/lib/tvgPicture.cpp
Comment (hermet) [inline src/lib/tvgPicture.cpp:126]: suggestion: !count -> count == 0

File: src/lib/tvgPictureImpl.h
Comment (hermet) [inline src/lib/tvgPictureImpl.h:203]: convention
if (xx) {
} else {
}

File: src/lib/tvgPictureImpl.h
Comment (hermet) [inline src/lib/tvgPictureImpl.h:264]: convention:
if (xx) {

File: src/lib/tvgPictureImpl.h
Comment (hermet) [inline src/lib/tvgPictureImpl.h:269]: convention:
if () {
} else {
}

File: src/lib/sw_engine/tvgSwRenderer.cpp
Comment (hermet) [inline src/lib/sw_engine/tvgSwRenderer.cpp:185]: Suggest clear name here. triangleCnt

**Review by hermet — CHANGES_REQUESTED** [review]
Overall good to go. Here some nit-picks I'd like to ask you check them.
If you have any opinions against them please comments. :)

**Comment by hermet** [comment]
One more, the commits are too messy here. If you clean up with the minimum necessary patch set. It would be much better.
Otherwise, I will squash them into one commit.

File: inc/thorvg.h
Comment (projectitis) [reply inc/thorvg.h:191]: @hermet I originally [matched skia](https://github.com/rive-app/rive-cpp/blob/master/skia/renderer/src/skia_factory.cpp#L170), which uses 0 - width/height, but I agree that 0 - 1 is better and I will change it.

File: inc/thorvg.h
Comment (projectitis) [reply inc/thorvg.h:1221]: @hermet I agree. Will change to be set only.

File: inc/thorvg.h
Comment (projectitis) [reply inc/thorvg.h:1224]: @hermet Agree. Will remove mesh if triangles==nullptr or count is 0.

File: src/lib/sw_engine/tvgSwCommon.h
Comment (projectitis) [reply src/lib/sw_engine/tvgSwCommon.h:324]: @hermet I agree - I actually just used count because I saw it used in other functions, like `Result colorStops(const ColorStop* colorStops, uint32_t cnt) noexcept;` but triangleCount is better.

**Comment by projectitis** [comment]
@hermet Thank you for the review. I have addressed all your comments. The biggest changes are:

1. Polygon (triangle) now expects normalized UV coordinates. In order to do this, the renderer converts them to image coordinates immediately before rendering. This is necessary because the user can load a new image (with same mesh) which might have a different size.
2. Picture.mesh will now clear the mesh if called with nullptr or count==0

Aside from that, updates are argument names, linting and comments (including @BETA_API tag).

I am happy to receive more suggestions.

Please squash commits when this PR is ready to merge.

**Review by JSUYA — APPROVED** [review]
Overall I think it's good. Thanks again for your work.
I left comments, please check it.
and I approved. Wait for @hermet's approval

File: src/lib/tvgPictureImpl.h
Comment (hermet) [inline src/lib/tvgPictureImpl.h:275]: I missed one point, duplicate() must copy the interal member data of the Picture. So do polygon data as well.
Could you please copy mesh data here as well?

**Review by hermet — CHANGES_REQUESTED** [review]
Please check comments one more, Also could you please rebase the patch so that we can merge patch without any conflicts?

File: src/lib/tvgPicture.cpp
Comment (projectitis) [reply src/lib/tvgPicture.cpp]: No longer valid (function has changed)

**Comment by projectitis** [comment]
@hermet Thanks for remembering the duplicate() function. I also double-checked all comments and fixed some additional linting. The only difference now that I can see from your comments is `triangleCnt` vs `triangleCount`.

I squashed commits and rebased to master.

Hopefully this is all good to go once the checks have passed.

**Review by hermet — APPROVED** [review]

**Comment by hermet** [comment]
applied with this: 
https://github.com/Samsung/thorvg/commit/3dd65dfed00849f0bd9d0bb0ade177fa961cd7a5

thanks!

**Pull request closed without merging**
